File: 2024/day04_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    matrix = [row.strip() for row in content[::]]
    return len(
        [
            True
            for row in range(len(matrix))
            for col in range(len(matrix[0]))
            if matrix[row][col] == "A" and check(matrix, row, col)
        ]
    )


def check(matrix: list[str], row: int, col: int):
    if row == 0 or row == len(matrix) - 1 or col == 0 or col == len(matrix[0]) - 1:
        return False
    return (
        (
            matrix[row - 1][col - 1] == "M"
            and matrix[row + 1][col + 1] == "S"
            and matrix[row + 1][col - 1] == "M"
            and matrix[row - 1][col + 1] == "S"
        )
        or (
            matrix[row - 1][col - 1] == "M"
            and matrix[row + 1][col + 1] == "S"
            and matrix[row - 1][col + 1] == "M"
            and matrix[row + 1][col - 1] == "S"
        )
        or (
            matrix[row + 1][col + 1] == "M"
            and matrix[row - 1][col - 1] == "S"
            and matrix[row - 1][col + 1] == "M"
            and matrix[row + 1][col - 1] == "S"
        )
        or (
            matrix[row + 1][col + 1] == "M"
            and matrix[row - 1][col - 1] == "S"
            and matrix[row + 1][col - 1] == "M"
            and matrix[row - 1][col + 1] == "S"
        )
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2024/day04-2.hs
solution :: String -> Int
solution content = length . filter (check matrix) $ [(x, y) | x <- [0 .. length matrix -1], y <- [0 .. hlength matrix -1], matrix !! x !! y == 'A']
    where matrix = lines content

check :: [String] -> (Int, Int) -> Bool
check matrix (row, col)
    | row == 0 || col == 0 || row == length matrix -1 || col == hlength matrix -1 = False
    | otherwise = any (\f -> f matrix (row, col)) [top, bottom, left, right]

top :: [String] -> (Int, Int) -> Bool
top matrix (row, col) = 
    matrix !! (row-1) !! (col-1) == 'M' &&
    matrix !! (row-1) !! (col+1) == 'M' &&
    matrix !! (row+1) !! (col-1) == 'S' &&
    matrix !! (row+1) !! (col+1) == 'S'

bottom :: [String] -> (Int, Int) -> Bool
bottom matrix (row, col) = 
    matrix !! (row+1) !! (col-1) == 'M' &&
    matrix !! (row+1) !! (col+1) == 'M' &&
    matrix !! (row-1) !! (col-1) == 'S' &&
    matrix !! (row-1) !! (col+1) == 'S'

left :: [String] -> (Int, Int) -> Bool
left matrix (row, col) = 
    matrix !! (row-1) !! (col-1) == 'M' &&
    matrix !! (row+1) !! (col-1) == 'M' &&
    matrix !! (row-1) !! (col+1) == 'S' &&
    matrix !! (row+1) !! (col+1) == 'S'

right :: [String] -> (Int, Int) -> Bool
right matrix (row, col) = 
    matrix !! (row-1) !! (col+1) == 'M' &&
    matrix !! (row+1) !! (col+1) == 'M' &&
    matrix !! (row-1) !! (col-1) == 'S' &&
    matrix !! (row+1) !! (col-1) == 'S'


hlength :: [[a]] -> Int
hlength = length . head

main :: IO ()
main = do readFile "day04.dat" >>= print . solution
